File: apps/android/core/src/main/java/com/aatech/betweenus/core/data/Markup.kt
package com.aatech.betweenus.core.data

/**
 * The markdown-ish shape of a message body.
 *
 * `MessageBody` decides what a message *is* - words, attachments and the quote
 * above a reply. This decides how the words read: the small set of marks every
 * chat app has trained people to type, and nothing else. There is no link
 * syntax, because a bare URL is already a link on every client, and no images,
 * because an attachment is not a body.
 *
 * Headings are the one thing that depends on what is being read. A heading in a
 * chat line is shouting, so [parse] does not have them; release notes are a
 * document and are nothing but headings, so [parseNotes] does. It is the same
 * parser with one rule switched on.
 *
 * Line for line the desktop client's `markup.ts`. Changing one changes both, or
 * the same message reads differently on two screens.
 *
 * Pure and offset-exact, which is the whole reason it is here rather than in
 * the renderer: the marks are *removed* from the text, so every span is an
 * index into the text that comes out, and the emoji splitter runs over that
 * same string afterwards. A parser that left the asterisks in would need no
 * offsets at all - and would draw them.
 */
object Markup {

    enum class Style { Bold, Italic, Strike, Code }

    /** A style over `[start, end)` of the block's own text. */
    data class Span(val start: Int, val end: Int, val style: Style)

    enum class Kind { Body, Quote, Code, Bullet, Number, Heading }

    /**
     * One block of a message.
     *
     * A list is not a block: every item is its own, and a run of them next to
     * each other is what a list *is*. That keeps this list flat - no tree, no
     * nesting - which is the whole of what a chat line needs and a fraction of
     * what a document parser would cost.
     *
     * [ordinal] is the number a `Number` item is drawn with, and the level of
     * a `Heading` - `## Install` is 2. Zero for everything else.
     */
    data class Block(
        val kind: Kind,
        val text: String,
        val spans: List<Span> = emptyList(),
        val ordinal: Int = 0,
    )

    private data class Delimiter(val token: String, val style: Style)

    // Longest first: `**` has to be tried before `*` or every bold is two
    // italics with nothing between them.
    private val DELIMITERS = listOf(
        Delimiter("**", Style.Bold),
        Delimiter("~~", Style.Strike),
        Delimiter("*", Style.Italic),
        Delimiter("_", Style.Italic),
        Delimiter("`", Style.Code),
    )

    private const val ESCAPABLE = "*_~`\\>"

    /**
     * A list marker, and the space after it.
     *
     * The space is what keeps `*bold*` from being a bullet: a marker glued to
     * its word is a mark, and a marker standing off from it is a list. The
     * indent is bounded at three so a deliberately indented line stays prose.
     */
    private val BULLET = Regex("^ {0,3}[-*+] +(.*)$")
    private val NUMBER = Regex("^ {0,3}(\\d{1,9})[.)] +(.*)$")

    /** `### Features`. The space is what keeps a `#channel` mention from being one. */
    private val HEADING = Regex("^ {0,3}(#{1,6}) +(.*?)\\s*#*\\s*$")

    /**
     * The `| --- | --- |` under a table's header row.
     *
     * ponytail: tables are not parsed, only this line is swallowed - the rows
     * draw as the pipes they are, which is legible for the three-row table a
     * release note carries. A real table block if one ever gets wide enough to
     * need it.
     */
    private val TABLE_RULE = Regex("^ {0,3}\\|[\\s:|-]*-[\\s:|-]*\\|\\s*$")

    /**
     * The same parser, reading a document rather than a chat line.
     *
     * Release notes are `### Features` and a list under it, which [parse] would
     * draw as a line beginning with three hashes. This is the only caller that
     * wants headings, and chat is the only one that must not have them.
     */
    fun parseNotes(text: String): List<Block> = parse(text, headings = true)

    fun parse(text: String, headings: Boolean = false): List<Block> {
        val blocks = mutableListOf<Block>()
        val lines = text.split("\n")
        var i = 0

        val pending = StringBuilder()
        var pendingQuote = false

        fun flush() {
            if (pending.isEmpty()) return
            val kind = if (pendingQuote) Kind.Quote else Kind.Body
            blocks += inline(pending.toString(), kind)
            pending.setLength(0)
            pendingQuote = false
        }

        while (i < lines.size) {
            val line = lines[i]

            if (line.trimStart().startsWith("```")) {
                flush()
                val fence = mutableListOf<String>()
                i++
                while (i < lines.size && !lines[i].trimStart().startsWith("```")) {
                    fence += lines[i]
                    i++
                }
                // An unclosed fence is still a code block. Somebody who opened
                // one and hit send meant the rest to be code; drawing it as
                // prose with three backticks in front of it helps nobody.
                if (i < lines.size) i++
                blocks += Block(Kind.Code, fence.joinToString("\n").trim('\n'))
                continue
            }

            if (headings) {
                if (TABLE_RULE.matches(line)) {
                    i++
                    continue
                }
                val heading = HEADING.find(line)
                if (heading != null) {
                    flush()
                    blocks += inline(
                        heading.groupValues[2],
                        Kind.Heading,
                        heading.groupValues[1].length,
                    )
                    i++
                    continue
                }
            }

            // A list item is its own block, so it ends whatever paragraph was
            // being gathered. Bullets first: `1.` cannot start with `-`, but a
            // bullet line can perfectly well contain a number.
            val bullet = BULLET.find(line)
            if (bullet != null) {
                flush()
                blocks += inline(bullet.groupValues[1], Kind.Bullet)
                i++
                continue
            }

            val numbered = NUMBER.find(line)
            if (numbered != null) {
                flush()
                // The run is renumbered from whatever the first item said, so
                // the `1. 1. 1.` everybody types comes out 1, 2, 3 - and a list
                // that deliberately starts at 5 still starts at 5.
                val previous = blocks.lastOrNull()
                val ordinal = if (previous != null && previous.kind == Kind.Number) {
                    previous.ordinal + 1
                } else {
                    numbered.groupValues[1].toIntOrNull() ?: 1
                }
                blocks += inline(numbered.groupValues[2], Kind.Number, ordinal)
                i++
                continue
            }

            val quoted = line.startsWith("> ") || line == ">"
            val body = if (quoted) line.removePrefix(">").removePrefix(" ") else line

            // A run of lines is one block only while they agree about being
            // quoted; the marker changing is a new block.
            if (pending.isNotEmpty() && quoted != pendingQuote) flush()
            if (pending.isNotEmpty()) pending.append('\n')
            pendingQuote = quoted
            pending.append(body)
            i++
        }
        flush()

        return blocks
    }

    private fun inline(source: String, kind: Kind, ordinal: Int = 0): Block {
        val out = StringBuilder()
        val spans = mutableListOf<Span>()
        scan(source, out, spans)
        return Block(kind, out.toString(), spans, ordinal)
    }

    private fun scan(source: String, out: StringBuilder, spans: MutableList<Span>) {
        var i = 0
        while (i < source.length) {
            val c = source[i]

            if (c == '\\' && i + 1 < source.length && source[i + 1] in ESCAPABLE) {
                out.append(source[i + 1])
                i += 2
                continue
            }

            val opener = openerAt(source, i)
            if (opener != null) {
                val from = i + opener.token.length
                val close = source.indexOf(opener.token, from)
                if (close > from) {
                    val start = out.length
                    val content = source.substring(from, close)
                    // Code is literal: a backtick span is where somebody puts
                    // the asterisks they do not want eaten.
                    if (opener.style == Style.Code) out.append(content)
                    else scan(content, out, spans)
                    spans += Span(start, out.length, opener.style)
                    i = close + opener.token.length
                    continue
                }
            }

            out.append(c)
            i++
        }
    }

    private fun openerAt(source: String, at: Int): Delimiter? {
        for (delimiter in DELIMITERS) {
            if (!source.startsWith(delimiter.token, at)) continue
            // `snake_case_names` are not two italics. An underscore only opens
            // when it is not glued to a word on its left, which is the rule
            // that keeps identifiers readable without a backtick round them.
            if (delimiter.token == "_" && at > 0 && source[at - 1].isLetterOrDigit()) continue
            return delimiter
        }
        return null
    }

    /** The box's text and where the caret sits in it, after a newline was typed. */
    data class Continuation(val text: String, val caret: Int)

    /**
     * What a newline typed inside a list should do.
     *
     * Null for an ordinary newline anywhere else - the caller then does
     * nothing and the box behaves as it always did. This is what makes a list
     * feel like a list while it is being written rather than only once it has
     * been sent: typing `- eggs` and pressing return offers the next bullet,
     * the way every editor anybody has used does.
     *
     * An empty item ends the list instead of offering another one. Without
     * that, the only way out of a list is to delete a marker somebody never
     * typed, which is the single thing that makes auto-continuation
     * infuriating rather than helpful.
     *
     * Line for line the desktop's `continueList`. Changing one changes both.
     */
    fun continueList(text: String, caret: Int): Continuation? {
        val from = text.lastIndexOf('\n', caret - 1) + 1
        val line = text.substring(from, caret)

        val bullet = BULLET.find(line)
        val numbered = if (bullet == null) NUMBER.find(line) else null
        if (bullet == null && numbered == null) return null

        // An empty item is somebody asking to stop. The marker goes, and with
        // it the item they were about to write.
        val content = bullet?.groupValues?.get(1) ?: numbered!!.groupValues[2]
        if (content.isBlank()) {
            return Continuation(text.substring(0, from) + text.substring(caret), from)
        }

        // The content is the tail of the line by construction, so whatever is
        // in front of it is exactly the marker - indent, bullet character and
        // spacing included. Rebuilding it from the parts would quietly
        // normalise all three.
        val prefix = line.substring(0, line.length - content.length)
        val marker = if (bullet != null) {
            prefix
        } else {
            val next = (numbered!!.groupValues[1].toIntOrNull() ?: 1) + 1
            prefix.replaceFirst(Regex("\\d{1,9}"), next.toString())
        }
        val inserted = "\n$marker"
        return Continuation(
            text.substring(0, caret) + inserted + text.substring(caret),
            caret + inserted.length,
        )
    }

    /** Whether anything at all would be drawn differently. Saves a rebuild. */
    fun isPlain(text: String): Boolean =
        text.none { it == '*' || it == '_' || it == '~' || it == '`' || it == '\\' } &&
            text.lineSequence().none {
                it.startsWith("> ") || it == ">" || BULLET.matches(it) || NUMBER.matches(it)
            }
}

File: apps/desktop/src/services/markup.ts
/**
 * The markdown-ish shape of a message body.
 *
 * `message-body.ts` decides what a message *is* - words, attachments and the
 * quote above a reply. This decides how the words read: the small set of marks
 * every chat app has trained people to type, and nothing else. There is no link
 * syntax, because a bare URL is already a link on every client; no headings,
 * because a heading in a chat line is shouting; and no images, because an
 * attachment is not a body.
 *
 * Pure and offset-exact, which is the whole reason it is here rather than in
 * the renderer: the marks are *removed* from the text, so every span is an
 * index into the text that comes out, and the emoji splitter runs over that
 * same string afterwards. A parser that left the asterisks in would need no
 * offsets at all - and would draw them.
 *
 * Line for line the android client's `Markup.kt`. Changing one changes both,
 * or the same message reads differently on two screens.
 */

export type Style = 'bold' | 'italic' | 'strike' | 'code';

/** A style over `[start, end)` of the block's own text. */
export interface Span {
  start: number;
  end: number;
  style: Style;
}

export type Kind = 'body' | 'quote' | 'code' | 'bullet' | 'number';

/**
 * One block of a message.
 *
 * A list is not a block: every item is its own, and a run of them next to each
 * other is what a list *is*. That keeps this list flat - no tree, no nesting -
 * which is the whole of what a chat line needs and a fraction of what a
 * document parser would cost.
 *
 * `ordinal` is the number a `number` item is drawn with, and zero for
 * everything else.
 */
export interface Block {
  kind: Kind;
  text: string;
  spans: Span[];
  ordinal: number;
}

interface Delimiter {
  token: string;
  style: Style;
}

// Longest first: `**` has to be tried before `*` or every bold is two italics
// with nothing between them.
const DELIMITERS: Delimiter[] = [
  { token: '**', style: 'bold' },
  { token: '~~', style: 'strike' },
  { token: '*', style: 'italic' },
  { token: '_', style: 'italic' },
  { token: '`', style: 'code' },
];

const ESCAPABLE = '*_~`\\>';

/**
 * A list marker, and the space after it.
 *
 * The space is what keeps `*bold*` from being a bullet: a marker glued to its
 * word is a mark, and a marker standing off from it is a list. The indent is
 * bounded at three so a deliberately indented line stays prose.
 */
const BULLET = /^ {0,3}[-*+] +(.*)$/;
const NUMBER = /^ {0,3}(\d{1,9})[.)] +(.*)$/;

export function parse(text: string): Block[] {
  const blocks: Block[] = [];
  const lines = text.split('\n');
  let i = 0;

  let pending: string[] = [];
  let pendingQuote = false;

  const flush = (): void => {
    if (pending.length === 0) return;
    blocks.push(inline(pending.join('\n'), pendingQuote ? 'quote' : 'body'));
    pending = [];
    pendingQuote = false;
  };

  while (i < lines.length) {
    const line = lines[i] ?? '';

    if (line.trimStart().startsWith('```')) {
      flush();
      const fence: string[] = [];
      i++;
      while (i < lines.length && !(lines[i] ?? '').trimStart().startsWith('```')) {
        fence.push(lines[i] ?? '');
        i++;
      }
      // An unclosed fence is still a code block. Somebody who opened one and
      // hit send meant the rest to be code; drawing it as prose with three
      // backticks in front of it helps nobody.
      if (i < lines.length) i++;
      blocks.push({ kind: 'code', text: trimNewlines(fence.join('\n')), spans: [], ordinal: 0 });
      continue;
    }

    // A list item is its own block, so it ends whatever paragraph was being
    // gathered. Bullets first: `1.` cannot start with `-`, but a bullet line
    // can perfectly well contain a number.
    const bullet = BULLET.exec(line);
    if (bullet) {
      flush();
      blocks.push(inline(bullet[1] ?? '', 'bullet'));
      i++;
      continue;
    }

    const numbered = NUMBER.exec(line);
    if (numbered) {
      flush();
      // The run is renumbered from whatever the first item said, so the
      // `1. 1. 1.` everybody types comes out 1, 2, 3 - and a list that
      // deliberately starts at 5 still starts at 5.
      const previous = blocks[blocks.length - 1];
      const ordinal =
        previous && previous.kind === 'number'
          ? previous.ordinal + 1
          : Number.parseInt(numbered[1] ?? '1', 10) || 1;
      blocks.push(inline(numbered[2] ?? '', 'number', ordinal));
      i++;
      continue;
    }

    const quoted = line.startsWith('> ') || line === '>';
    const body = quoted ? line.replace(/^>( )?/, '') : line;

    // A run of lines is one block only while they agree about being quoted;
    // the marker changing is a new block.
    if (pending.length > 0 && quoted !== pendingQuote) flush();
    pendingQuote = quoted;
    pending.push(body);
    i++;
  }
  flush();

  return blocks;
}

/** `trim('\n')`, which JavaScript has no single call for. */
function trimNewlines(text: string): string {
  return text.replace(/^\n+/, '').replace(/\n+$/, '');
}

function inline(source: string, kind: Kind, ordinal = 0): Block {
  const out = { text: '' };
  const spans: Span[] = [];
  scan(source, out, spans);
  return { kind, text: out.text, spans, ordinal };
}

/**
 * `out` is a holder rather than a plain string because a span's offsets are
 * measured against everything written so far, including by the recursive call
 * that handles a nested style - so the writer has to be shared, not copied.
 */
function scan(source: string, out: { text: string }, spans: Span[]): void {
  let i = 0;

  while (i < source.length) {
    const c = source[i] ?? '';

    if (c === '\\' && i + 1 < source.length && ESCAPABLE.includes(source[i + 1] ?? '')) {
      out.text += source[i + 1] ?? '';
      i += 2;
      continue;
    }

    const opener = openerAt(source, i);
    if (opener) {
      const from = i + opener.token.length;
      const close = source.indexOf(opener.token, from);
      if (close > from) {
        const start = out.text.length;
        const content = source.slice(from, close);
        // Code is literal: a backtick span is where somebody puts the
        // asterisks they do not want eaten.
        if (opener.style === 'code') out.text += content;
        else scan(content, out, spans);
        spans.push({ start, end: out.text.length, style: opener.style });
        i = close + opener.token.length;
        continue;
      }
    }

    out.text += c;
    i++;
  }
}

function openerAt(source: string, at: number): Delimiter | null {
  for (const delimiter of DELIMITERS) {
    if (!source.startsWith(delimiter.token, at)) continue;
    // `snake_case_names` are not two italics. An underscore only opens when it
    // is not glued to a word on its left, which is the rule that keeps
    // identifiers readable without a backtick round them.
    if (delimiter.token === '_' && at > 0 && /[\p{L}\p{N}]/u.test(source[at - 1] ?? '')) continue;
    return delimiter;
  }
  return null;
}

/** The box's text and where the caret sits in it, after a newline was typed. */
export interface Continuation {
  text: string;
  caret: number;
}

/**
 * What a newline typed inside a list should do.
 *
 * Null for an ordinary newline anywhere else - the caller then does nothing
 * and the box behaves as it always did. This is what makes a list feel like a
 * list while it is being written rather than only once it has been sent:
 * typing `- eggs` and pressing return offers the next bullet, the way every
 * editor anybody has used does.
 *
 * An empty item ends the list instead of offering another one. Without that,
 * the only way out of a list is to delete a marker somebody never typed, which
 * is the single thing that makes auto-continuation infuriating rather than
 * helpful.
 *
 * Pure, and separate from the two composers that call it, because the rule is
 * fiddly and identical on both.
 */
export function continueList(text: string, caret: number): Continuation | null {
  const from = text.lastIndexOf('\n', caret - 1) + 1;
  const line = text.slice(from, caret);

  const bullet = BULLET.exec(line);
  const numbered = bullet ? null : NUMBER.exec(line);
  if (!bullet && !numbered) return null;

  // An empty item is somebody asking to stop. The marker goes, and with it
  // the item they were about to write.
  const content = bullet ? bullet[1] : numbered![2];
  if ((content ?? '').trim().length === 0) {
    const without = text.slice(0, from) + text.slice(caret);
    return { text: without, caret: from };
  }

  // The content is the tail of the line by construction, so whatever is in
  // front of it is exactly the marker - indent, bullet character and spacing
  // included. Rebuilding it from the parts would quietly normalise all three.
  const prefix = line.slice(0, line.length - (content ?? '').length);
  const marker = bullet
    ? prefix
    : prefix.replace(/\d{1,9}/, String((Number.parseInt(numbered![1] ?? '1', 10) || 1) + 1));
  const inserted = `\n${marker}`;
  return {
    text: text.slice(0, caret) + inserted + text.slice(caret),
    caret: caret + inserted.length,
  };
}

/** Whether anything at all would be drawn differently. Saves a rebuild. */
export function isPlain(text: string): boolean {
  if (/[*_~`\\]/.test(text)) return false;
  return text
    .split('\n')
    .every((line) => !(line.startsWith('> ') || line === '>' || BULLET.test(line) || NUMBER.test(line)));
}
